### ISSUE TITLE
Add options from Weimer2018 especially errors for format-security and implicit-function-declaration

### ISSUE THREAD

**Issue by david-a-wheeler** [ossf/wg-best-practices-os-developers#342]
[Florian Weimer's 2018 article "Recommended compiler and linker flags for GCC"](https://developers.redhat.com/blog/2018/03/21/compiler-and-linker-flags-gcc) has a number of interesting ideas. We should review it for options.

At the very least, I like adding these options:

~~~ sh
-Werror=format-security \
-Werror=implicit-function-declaration
~~~

As already noted in the guide, adding a general `-Werror` for *all* warnings has a lot of negatives, as it makes compilation very dependent on a specific version of a specific compiler. However, adding *specific* warnings for *specific* cases does *not* have this downside. We should probably do more of it.

In this case, these turn "should never do that" cases into outright errors. Raising an error for implicit function declaration is also part of #339. I worry that autoconf might have some implicit function declarations (at least for exit()) - we may need to determine how to deal with that and document it.

Possible other options, though I'm not as confident:

~~~ sh
-fasynchronous-unwind-tables \
-fexceptions
~~~

Some discussion about implicit function breakage is here:
https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213

**Comment by david-a-wheeler** [comment]
Looks like the latest version of autoconf will address some of these. It's planned to be released next week (December 2023). A discussion about this should soon show here: https://lists.gnu.org/archive/html/autoconf/2023-12/threads.html

**Comment by thomasnyman** [comment]
I've looked into both `-fasynchronous-unwind-tables` and `-fexceptions` and from what I can tell the benefits of these are limited to improving precision of unwind information and improving the performance of profilers at the cost of larger binary sizes. `-fexceptions` may need to be enabled for C code that needs to interoperate with C++. I have not found anything indicating these options would have a security benefit, so I've proposed #379 adding these to the list of considered options.

**Comment by thomasnyman** [comment]
After digging further `-fexceptions`  is referenced by @fweimer in this [thread on the libc-alpha mailing list](https://sourceware.org/pipermail/libc-alpha/2017-November/088474.html) in relation to an issue which seems to match the description of unprotected function pointers being spilled by `pthread_cleanup_push` in the [Red Hat developers article](https://developers.redhat.com/blog/2018/03/21/compiler-and-linker-flags-gcc).

If I understand the context correctly, `-exceptions` is proposed as workaround for a [particular GCC bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61118) in the `-Wclobbered` option that is causing a spurious warning in the codepath of the thread cancellation handlers not dependent on C++ exception handling, but instead uses `setjmp` / `longjmp` for stack unwinding. The [fix to the warning in glibc](https://sourceware.org/pipermail/libc-alpha/2017-November/088469.html) is introducing a function pointer spill instead of GCC generating a direct call before the fix, so the security rationale is that relying on C++ exception handling is preferable to spilling that register.

The GCC bug report has since been merged to [Bug 2116](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=21161) which is still open. Ideally, once the GCC bug is fixed, glibc could in principle fall back to the code giving a direct call within the handler but, in the meantime, having the thread cancellations handlers use C++ exception handling is preferable, even when compiling (multi-threaded) C code.

In summary `-exceptions` benefits multi-threaded C code when utilizing glibc to avoid the drawback of a particular unideal code fix. I've put #379 back to draft mode to be reworked accordingly.

I still interpret `-fasynchronous-unwind-tables` as being orthogonal to the above, but further input on that would be welcome. 

It should be noted that C++ exception handling is also susceptible to [different types of target confusion conditions](https://www.ndss-symposium.org/wp-content/uploads/2023/02/ndss2023_s295_paper.pdf) if unwinding operates on corrupt stack frames. The mitigations to such attacks recommended by the researchers include extending stack-protector heuristics to cover any function that might throw.

**Comment by thomasnyman** [comment]
Thanks @fweimer. I had missed Jeff Law's explanation. I hope the [text we merged into the guide](https://best.openssf.org/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C%2B%2B#enable-exception-propagation-to-harden-multi-threaded-c-code) is still accurate. We did not reference `-Wclobbered` directly there.
